duplicate.py

```diff
@@ -0,0 +1,165 @@
+    feature_cleanup.responses(),
+    feature_cleanup.hex(),
+    feature_cleanup.dll(),
+    feature_cleanup.fileref(),
+    feature_cleanup.url(),
+    feature_cleanup.synonyms(),
+    feature_cleanup.crash(),
+]
+
+# A map from bug id to its duplicate ids
+duplicates = defaultdict(set)
+all_ids = set(
+    bug["id"]
+    for bug in bugzilla.get_bugs()
+    if bug["creator"] not in REPORTERS_TO_IGNORE and "dupeme" not in bug["keywords"]
+)
+
+for bug in bugzilla.get_bugs():
+    dupes = [entry for entry in bug["duplicates"] if entry in all_ids]
+    if bug["dupe_of"] in all_ids:
+        dupes.append(bug["dupe_of"])
+
+    duplicates[bug["id"]].update(dupes)
+    for dupe in dupes:
+        duplicates[dupe].add(bug["id"])
+
+
+class SimilarityLsi:
+    def __init__(self):
+
+        self.corpus = []
+        texts = []
+        self.ps = PorterStemmer()
+
+        for bug in bugzilla.get_bugs():
+            textual_features = bug["summary"] + " " + bug["comments"][0]["text"]
+            for func in cleanup_functions:
+                textual_features = func(textual_features)
+
+            # cleaning the texts
+            textual_features = re.sub("[^a-zA-Z0-9]", " ", textual_features)
+            self.corpus.append([bug["id"], textual_features])
+
+            # create bag of words
+            texts.append(
+                [
+                    self.ps.stem(word)
+                    for word in textual_features.lower().split()
+                    if word not in set(stopwords.words("english")) and len(word) > 1
+                ]
+            )
+
+        # Assigning unique integer ids to all words
+        self.dictionary = Dictionary(texts)
+
+        # conversion to bow
+        corpus_final = [self.dictionary.doc2bow(text) for text in texts]
+
+        # initializing and applying the tfidf transformation model on same corpus,resultant corpus is of same dimensions
+        tfidf = models.TfidfModel(corpus_final)
+        corpus_tfidf = tfidf[corpus_final]
+
+        # transform tfidf corpus to latent 300-D space via LATENT SEMANTIC INDEXING
+        self.lsi = models.LsiModel(
+            corpus_tfidf, id2word=self.dictionary, num_topics=300
+        )
+        corpus_lsi = self.lsi[corpus_tfidf]
+
+        # indexing the corpus
+        self.index = similarities.Similarity(
+            output_prefix="simdata.shdat", corpus=corpus_lsi, num_features=300
+        )
+
+    def get_similar_bugs(self, query, default=10):
+
+        # consider first comment too
+        query_summary = query["summary"] + " " + query["comments"][0]["text"]
+        for func in cleanup_functions:
+            query_summary = func(query_summary)
+
+        # cleaning the texts
+        query_summary = re.sub("[^a-zA-Z0-9]", " ", query_summary)
+        query_summary = [
+            self.ps.stem(word)
+            for word in query_summary.lower().split()
+            if word not in set(stopwords.words("english")) and len(word) > 1
+        ]
```

Comment:
There is some common code here and in __init__ that you could refactor in a separate function.